### PR TITLE
fix(ci): bake NEXT_PUBLIC_* into Capgo OTA bundles (native QR scan/paste dead on ota-1.0.54/1.0.55)

### PR DESCRIPTION
**Pull request by abalinda**
## Summary

`.github/workflows/capgo-deploy.yml` ran `node scripts/native-build.js` without first writing `.env.production.local`. `ios-release.yml` and `android-release.yml` write that file (every production `NEXT_PUBLIC_*`) right before the same build step; the OTA lane never did, since the lane was created on 2026-04-22.

So every production OTA bundle was built with each `process.env.NEXT_PUBLIC_*` set to `undefined`:

- `src/components/Global/DirectSendQR/utils.ts` declared `export const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL!`, and `recognizeQr()` read `BASE_URL.toLowerCase()` on its first line. Every payload — Pix, Mercado Pago, EVM, paste — threw a `TypeError` before any match ran, and the catch-all showed "Erro ao processar o código QR".
- `NEXT_PUBLIC_SENTRY_DSN` and `NEXT_PUBLIC_POSTHOG_KEY` were also undefined, so Sentry and PostHog never initialised. The error catcher ran; the SDKs were absent. That is why the failure left no trace anywhere.

`ota-1.0.54` and `ota-1.0.55` are the first two OTAs devices ever applied, so this surfaced only this week.

Note the shape, because it is what the guards below are built around: **the store binary is fine** — both native lanes bake env correctly — and **every OTA that lands on top of it is not**. The app works on install and breaks the first time an OTA applies.

Full RCA: mono `ops/native-ota-envless-bundle-rca.md`.

## What this PR does

**Closes the hole**

1. `capgo-deploy.yml` gets the `.env.production.local` step, identical to the one in the two store lanes.
2. All three lanes write that file with a **quoted** heredoc delimiter. Unquoted, a secret value containing `$` or a backtick was expanded — and command-substituted — by the shell before reaching the file. Demonstrated locally: a value of ``https://ex.io/p$roject/`id -un`/x`` reached the file as `https://ex.io/p/<output of id -un>/x` — `$roject` silently dropped, and the backticks executed on the runner. `${{ }}` is interpolated by the runner before bash sees the script, so it still substitutes.

**Two guards, one per stage**

3. **Before the build**, `native-build.js` fails closed under `CI`: a missing file, or a required key with no value, throws instead of warning. Locally it still warns, so `pnpm native:build` on a dev machine is unchanged (rpId stays fatal everywhere — passkeys break without it).
4. **After the build**, it scans `out/` for each required value and fails if one left no trace. The first guard can only see the file; this one proves next actually inlined it. In `native-build.js` rather than the workflow, so all three lanes get it.

The file is read with **dotenv**, the parser `@next/env` uses to bake it, rather than a per-key regex. A regex disagrees with dotenv in three ways that each end in a bundle the check called healthy:

| `.env.production.local` contains | a per-key regex says | what next bakes |
|---|---|---|
| `KEY=value` then later `KEY=` | present ✅ | `""` (dotenv keeps the **last**) |
| `KEY= # todo` | present, value `# todo` | `""` (dotenv strips inline comments) |
| `export KEY=value` | **missing** ❌ | `"value"` (dotenv accepts `export `) |

**Stops the remaining silent-undefined reads**

5. `DirectSendQR/utils.ts` drops its duplicate `BASE_URL` and imports the one from `@/constants/general.consts`, which falls back to `https://peanut.me`. Nothing imported the local copy.
6. `useCreateLink` built the claim URL from `process.env.NEXT_PUBLIC_BASE_URL!` — in an env-less bundle that handed users links reading `undefined/claim`, with nothing thrown. Now the same `general.consts` `BASE_URL`.
7. `BUNDLER_URL` / `PAYMASTER_URL` read `undefined`, and viem's `http()` takes that as "use the chain's public RPC" — which has no ERC-4337 methods, so every userOp failed with an error naming neither ZeroDev nor the missing variable. `assertZeroDevRpcUrls()` now runs at the three functions that build the transports.
8. The OneSignal web adapter already threw on a missing app id and safari id; the webhook url had a `!` and was left out of that check. It joins it.

## Task

TASK-21900 — https://app.notion.com/3c883811757981ff876fc1bcdf1e88c3

## Risks / breaking changes

- Every future `ota-*` tag now ships a bundle with env baked. Sentry, PostHog and ZeroDev go live on OTA for the first time — expect native events and errors to appear that were previously dark.
- The `dev` → staging lane in the same workflow gets the same production values. That is not a redirect: `NEXT_PUBLIC_BASE_URL` and `NEXT_PUBLIC_PEANUT_API_URL` already resolved to their production defaults through code fallbacks in the env-less bundle.
- Either guard turns a lane that forgets the env file into a red build. Intended — that is the whole point.
- **`assertZeroDevRpcUrls()` adds a throw to the wallet path.** It is deliberately at the three transport-building functions and *not* at module load: a wallet-only misconfiguration must not white-screen the app, and it must not break a local or preview build on import. In `createEphemeralSpendSession` it sits above the `try`, because that `catch` rewraps everything into `EphemeralKeyPreflightError` and would otherwise rebury the message as a `cause`. It cannot fire when the env is baked, which after this PR is enforced twice.
- `.env.test` gains the two ZeroDev URLs, so a suite that reaches the kernel client keeps testing what it meant to instead of failing on the new guard.
- `dotenv` is added as a devDependency, pinned `^16.6.1` rather than the newest 17.4.2. `parse()` is byte-identical between the two, 16.6.1 is already in the lockfile (so the entry dedupes to 3 lines), and `jest.setup.ts`'s existing `require('dotenv')` — until now an undeclared dependency — keeps resolving to the version it already resolved to, instead of gaining dotenv 17's per-suite log line.
- No runtime change on web: `general.consts` `BASE_URL` is the value the web bundle already used.
- No cross-repo impact. No API change.

## QA

- **CI**: the `capgo-deploy` run log must show `✅ NEXT_PUBLIC_NATIVE_RP_ID=peanut.me`, `✅ <KEY> is set` for the other seven, and `✅ all 8 required NEXT_PUBLIC_* found in the exported bundle`. Both OTA runs to date showed `⚠️  .env.production.local not found`.
- **Device**: after a human tags `ota-1.0.56`, kill and reopen the app twice (launch 1 downloads, launch 2 applies), then scan a Pix QR — it must resolve. Sentry `environment:native release:<7-char sha>` canaries appearing is the proof the bundle is live and not dark.
- **Unit**: `scripts/__tests__/native-env-check.test.js` covers the required-key check (all present, empty value, absent key, quoted-empty, commented-out, repeated key, inline comment, `export ` prefix) and the bundle scan (all inlined, one absent, a key with no value to look for, a value hiding in a non-`.js` asset). The three parser cases each fail against the pre-review implementation. The case in `recognizeQr.test.ts` deletes `NEXT_PUBLIC_BASE_URL`, reloads the module, and asserts a Pix payload and a `peanut.me` URL still resolve — it fails with the exact production `TypeError` against the pre-fix file.
- Local gate: `prettier --check`, `tsc --noEmit`, `eslint` on the changed files, and the 18 suites covering the wallet, withdraw, send-link, request and add-money paths the new guards sit in — all pass. Full `unit` job green in CI.

### Measured, not assumed

- **Parity with the real parser.** The check and `@next/env@16.2.3`'s own `loadEnvConfig` agree on all six env-file shapes, including the three in the table above.
- **What actually gets inlined.** A full native build with sentinel values, run on **both `main` and `dev`**, puts all 8 required values in `out/**/*.js` and nowhere else — which is why only `.js` is scanned. `NEXT_PUBLIC_ZERO_DEV_PASSKEY_PROJECT_ID` and `NEXT_PUBLIC_GA_KEY` reach no shipped module at all, which is why neither is on the required list. Adding a key to that list without measuring it first would red-build the release lane, so the comment beside it says so.
- **Cost of the scan**: ~100ms over the 51 MB / 408-file bundle. Against a deliberately drifted env it reported exactly the two drifted keys.

## Screenshots

N/A (no visible change).

## Design notes / accepted trade-offs

The env block is still duplicated across three workflows on purpose. The durable dedupe is one reusable `workflow_call` job (or a composite action) with `secrets: inherit`, and that is deferred to a follow-up on `dev` — #2806 rewrote the ios/android release lanes and #2825 rewrites all three again, so doing it here guarantees a back-merge conflict on exactly these files. The two guards in `native-build.js` are what make the duplication safe meanwhile: a lane that drifts breaks its own build instead of shipping a dark bundle.

**Still not a durable fix on its own.** This PR closes the exact hole and makes a misbuilt bundle fail its own build twice over. It does not make OTA releases safe in general. Follow-up, planned right behind this one:
- App refuses a misbuilt bundle at runtime: gate `CapacitorUpdater.notifyAppReady()` on a self-check (baked env present + `/healthz`), so Capgo auto-rolls back within 15s on every device.
- One source for the native env (`workflow_call` build job with `secrets: inherit`) and a required-key list derived from code, not hand-kept.
- Telemetry that cannot be stripped by env: public DSN / PostHog key as code defaults, `x-peanut-client` bundle header on native API calls, Sentry tags `capgo_bundle` / `native_version`, alert on native canary users/day dropping.
- Bundle ↔ shell fingerprint check in `capgo-deploy.yml`, staged rollout, and a post-OTA "did the new release phone home" check.

## Rollout

1. Merge this to `main`.
2. A human tags `ota-1.0.56` on the merge commit and pushes the tag. The workflow runs from the tagged commit, so the fix has to be on it.
3. Check the run log and Capgo, then verify on device.
4. Back-merge `main` → `dev`.

**Order matters against #2825.** That PR replaces tag-driven releases with a "Release OTA" `workflow_dispatch` pinned to `dev`, resolving `<major>.<build>.<ota>` instead of taking the number from the tag. Run against the repo's real tag list today, its resolver refuses both paths — `ota` errors with *"no v1.\<build\>.0 tag exists yet — cut a native release before an OTA"* (only `v1.0.0` exists), and the break-glass tag `ota-1.0.56` fails validation with *"has build 0 — build numbers start at 1"*. So if #2825 lands first, this hotfix cannot ship as an OTA without first cutting a full `1.1.0` native store release. Ship steps 1–3 before merging #2825.

The two branches otherwise compose cleanly: trial-merged, they auto-merge on all three workflows, with the env step landing inside #2825's restructured `deploy` job right before `native-build.js` and the quoted delimiters surviving in all three files.

**On the back-merge (step 4).** `src/config/peanut.config.tsx` conflicts between `main` and `dev` today, with or without this PR — this PR adds zero new conflicts. It needs a real hand-merge (`dev` has the `AppStateProviders` dynamic-import + marketing-route split, `main` has the passkey-ceremony shim + native canary; both belong in the result). Do not resolve it by copying one side's hunk wholesale — git will not recognise it as the same change and silently duplicates the block, which `merge-tree` reports as exit 0. Signatures are clear: 0 unsigned commits in `origin/dev..origin/main` and GitHub reports all 24 as `verified: true`, so the merge will not hit `GH013` — provided the branch is cut from that day's `origin/dev`.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved QR code recognition for Peanut payment payloads and URLs when the native app’s base URL configuration is unavailable.
  - Preserved special characters in production secrets during native and over-the-air deployments.
  - Native builds now detect missing or unbundled production configuration earlier.

- **Reliability**
  - Added validation for required blockchain bundler and paymaster settings.
  - Improved deployment checks for API, authentication, analytics, monitoring, and notification configuration.
  - Added notification webhook validation to improve release readiness.

### CODE REVIEW

**Comment by abalinda** [comment]
@coderabbitai review

File: src/components/Global/DirectSendQR/__tests__/recognizeQr.test.ts [hunk 1]
```diff
@@ -632,4 +632,32 @@ describe('recognizeQr', () => {
...
+    // local `BASE_URL = process.env.NEXT_PUBLIC_BASE_URL!` was read on the first line
+    // of recognizeQr, so every payload threw a TypeError before any match ran — mono
+    // `ops/native-ota-envless-bundle-rca.md`. BASE_URL now comes from
+    // @/constants/general.consts, which falls back to https://peanut.me.
+    describe('with no NEXT_PUBLIC_BASE_URL (env-less native bundle)', () => {
+        const original = process.env.NEXT_PUBLIC_BASE_URL
+
+        afterEach(() => {
+            process.env.NEXT_PUBLIC_BASE_URL = original
+        })
```
[the remark this answers is not in the document]
Comment (abalinda) [reply src/components/Global/DirectSendQR/__tests__/recognizeQr.test.ts:648]: Fixed in 7095b7682 — afterEach now deletes the key when it was originally absent instead of assigning undefined.

File: scripts/native-build.js [hunk 2]
```diff
@@ -477,6 +477,34 @@ function restoreModifiedFiles() {
...
+    'NEXT_PUBLIC_ZERO_DEV_PAYMASTER_URL',
+    // the native OneSignal adapter throws at init without it — push dead on every device
+    'NEXT_PUBLIC_ONESIGNAL_APP_ID',
+]
+
+// Value of a `KEY=value` line, or '' when the key has no line of its own or the
+// line carries no value (`KEY=`). Anchoring per line also skips comments.
+function nativeEnvValue(envContent, key) {
+    const match = envContent.match(new RegExp(`^${key}=(.*)$`, 'm'))
+    return match ? match[1].trim() : ''
```
[the remark this answers is not in the document]
Comment (abalinda) [reply scripts/native-build.js]: Fixed in d1aaee218 — nativeEnvValue() now unwraps matching single/double quotes before trimming, so KEY="" and KEY='  ' count as missing; test added for quoted-empty, quoted-whitespace and quoted-set values.

**Review by innolope-dev — APPROVED** [review]
Approved. I authored the last two commits though (61243785d, 3a95199b1).

Verified before merging:
- 13/13 checks green; no unresolved review threads.
- The env check agrees with @next/env@16.2.3's own loadEnvConfig on all six env-file shapes, including the three a per-key regex got wrong (repeated key, inline comment, `export ` prefix).
- Sentinel native builds on both main and dev put all 8 required NEXT_PUBLIC_* in out/**/*.js, so the post-build scan will not false-fail either lane after the back-merge.
- main -> dev trial-merged: this PR adds zero new conflicts (peanut.config.tsx already conflicts between the two branches), and there are 0 unsigned commits in origin/dev..origin/main with all 24 GitHub-verified, so the back-merge will not hit GH013.
